BUG: prevent ausearch race conditions

WOnder93 opened the pull request:
On busy machines it can happen that the record we are searching for
doesn't make it to the log file before we run ausearch.

This patch fixes this problem by adding an 'audit_sync' script, which
generates a marker user record using 'auditclt -m' and waits until the
marker appears in the audit log. This is much more reliable than the
current mix of either completely ignoring the issue or just doing an
unreliable sleep.

--
We are observing some rare random failures in our infrastructure and this seems to be the most logical explanation for them. The failures are hard to trigger, so I wasn't able to verify if this patch really fixes the problem. However, it is easy to see that the race *can* happen given the right circumstances and the solution is IMHO simple enough that it is worth applying it.

rgbriggs commented:
On 2019-04-24 06:14, Ondrej Mosnáček wrote:
> This patch fixes this problem by adding an 'audit_sync' script, which generates a marker user record using 'auditclt -m' and waits until the marker appears in the audit log. This is much more reliable than the current mix of either completely ignoring the issue or just doing an unreliable sleep.

This looks like a helpful idea.

Might it make sense to move the duplicated "key_gen()" function from each test into a function library, and use this to generate a marker as well rather than coming up with yet another random nonce for this service?

Can we avoid the basedir code?

WOnder93 commented:
> Might it make sense to move the duplicated "key_gen()" function from each test into a function library, and use this to generate a marker as well rather than coming up with yet another random nonce for this service?

From what I heard from @The-Mule, @pcmoore prefers the tests to be as self-contained as possible, so I think that's the reason why the functions are defined in-line now. I'll leave it up to @pcmoore to decide about this. I think that key_gen() is simple enough that having it duplicated many times is tolerable.

Of course, this patch already breaks that self-containment, but I don't see a better way to address the problem. Also note that even if the `audit_sync` script is not found, the tests will still work as before the patch (although stderr will be cluttered).

One possible alternative would be to adjust PATH in the Makefile so that a wrapper script around the `ausearch` binary is used when the `ausearch` command is invoked from the tests. The downside of this is that we would now do "audit_sync" before every `ausearch` call, even when it is not logically needed.

> Can we avoid the basedir code?

Not if we want to make sure that the test scripts can be run from an arbitrary CWD (as it is now).

The-Mule commented:
Yes, I think we definitely need to address issues mentioned in the description. When testing on non-Intel machines sleeps become very unreliable and we are hitting a lot of false positives. Test cases should be compatible with a very slow machine and we needs some kind of "active waiting" with a specific time-out. That is exactly what Ondrej provided I would say.

 I also agree that there are benefits of keeping test cases self-contained (ie. avoid using library for now). As Ondrej pointed out, having a small chunks of code duplicated in each test is not a burden. Actually this "waiting" code is the first think which would justify a library.

pcmoore commented:
> From what I heard from @The-Mule, @pcmoore prefers the tests to be as self-contained as possible, so I think that's the reason why the functions are defined in-line now. I'll leave it up to @pcmoore to decide about this.

Yes, I would like to keep the tests self-contained for as long as possible.  We may need to break that at some point, but this problem/solution doesn't appear to be enough of a reason to lose this.

I've got some additional thoughts on this PR, but I'll comment inline.

pcmoore commented on tests/audit_sync:
```diff
@@ -0,0 +1,15 @@
+
+auditctl -m "$key"
+
+wait $!
```
I'm not in love with the idea that this is a separate script that lives outside the individual tests for reasons already discussed in the PR comments (self contained tests).

This also seems overly complex, why couldn't we do something like this in the tests?

```
# 
# untested perl-ish pseudo code, probably buggy/flawed as a result
#

system("auditctl -m <somerandomstring>");
while (system("ausearch -m USER | grep -q msg=<string>")) {
     # sleep and/or sync, possibly a loop counter and a fatal error to prevent excessive looping
}
```

WOnder93 replied:
For performance reasons - every time you call ausearch, it goes through the whole audit log (including rotated logs). I tried to minimize the number of times the logs need to be traversed. Also, the script doesn't use the somewhat slower ausearch (which parses each record), but ony a simple grep.

pcmoore replied:
At this point in time I'm not overly concerned about the overhead of an additional ausearch call for each test; the test suite runs rather quickly as it currently stands.  If you are bothered by the ausearch call you could easily substitute a call to sed/grep.

pcmoore replied:
I want to add, the important part about this comment thread is not the tool that is used to search through the logs for the marker, but rather the idea that this explicit check (vs an implicit sleep) is incorporated into the tests and not a separate script.

WOnder93 commented:
OK, I rewrote the patch to use inline snippets for the syncing and switched to the simpler "ausearch | grep" logic. (And I also found and removed some `$basedir` declarations that were not used.)

pcmoore commented:
Merged via f7023d428cc54eeff48c0f97f69fbdde0a043d1a, thanks!
